src/basic/mkdir.c

```diff
@@ -31,27 +32,27 @@ int mkdir_safe_internal(
         assert(mode != MODE_INVALID);
         assert(_mkdirat && _mkdirat != mkdirat);
 
-        if (_mkdirat(AT_FDCWD, path, mode) >= 0) {
-                r = chmod_and_chown(path, mode, uid, gid);
+        if (_mkdirat(dir_fd, path, mode) >= 0) {
+                r = chmod_and_chown_at(dir_fd, path, mode, uid, gid);
                 if (r < 0)
                         return r;
         }
 
-        if (lstat(path, &st) < 0)
+        if (fstatat(dir_fd, path, &st, AT_SYMLINK_NOFOLLOW) < 0)
                 return -errno;
 
         if ((flags & MKDIR_FOLLOW_SYMLINK) && S_ISLNK(st.st_mode)) {
                 _cleanup_free_ char *p = NULL;
 
-                r = chase_symlinks(path, NULL, CHASE_NONEXISTENT, &p, NULL);
+                r = chase_symlinks_at(dir_fd, path, CHASE_NONEXISTENT, &p, NULL);
```

Comment:
hmm, so this looks wrong. the second argument to chase_symlinks() was the root filesystem we take absolute symlinks relative to. But in your new code you actually take them relative to dir_fd?

Reply:
So it's kinda complicated. The best way to solve this would be to have chase_symlinks_at() take both a root directory fd and a separate directory fd. But to get that to work, we need a way to transform the path relative to the directory fd to a path relative to the root directory fd. The only way I know of to do that is via /proc. If that's viable, it should be possible to add chase_symlinks_at_full() that takes a root directory fd, a directory fd and a path relative to the directory fd, transform the path to be relative to the root directory fd via /proc and then call chase_symlinks_at() with the transformed path and the root directory fd. 

I'm not sure if the current behavior makes sense either though. If we look at the usages of `mkdir_parents_safe()` and `mkdir_p_safe()` in nspawn-mount.c, they pass a root directory as the prefix, but the symlinks are still resolved against the host's root directory, that doesn't seem correct to me.